packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -166,120 +169,201 @@ class TimelineData {
+    if (event.isCpuEvent) {
+      if (_cpuDurationStack != null) {
+        _cpuDurationStack.addChild(e, findChildLocation: true);
+      } else {
+        _maybeAddEvent(e);
+      }
+    }
+    if (event.isGpuEvent) {
+      if (_gpuDurationStack != null) {
+        _gpuDurationStack.addChild(e, findChildLocation: true);
+      } else {
+        _maybeAddEvent(e);
+      }
     }
   }
 
-  void _handleAsyncBeginEvent(TraceEvent event) {
-    final String asyncUID = event.asyncUID;
-
-    final TimelineEvent parentEvent = _asyncEvents[asyncUID];
-    if (parentEvent == null) {
-      final TimelineEvent e = TimelineEvent(
-        event.name,
-        event.timestampMicros,
-        _getTimelineEventType(event),
-      );
-      _asyncEvents[asyncUID] = e;
-    } else {
-      final TimelineEvent e = TimelineEvent(
-        event.name,
-        event.timestampMicros,
-        _getTimelineEventType(event),
-      );
-      e.parent = parentEvent;
-      _asyncEvents[asyncUID] = e;
+  /// Looks through [_pendingEvents] and attempts to add events to frames in
+  /// [_pendingFrames].
+  void _maybeAddPendingEvents() {
+    final List<TimelineFrame> frames = _getAndSortWellFormedFrames();
+    for (TimelineFrame frame in frames) {
+      // Sort _pendingEvents by their startTime. This ensures we will add the
+      // first matching event within the time boundary to the frame.
+      _pendingEvents.sort((a, b) => a.startTime.compareTo(b.startTime));
+
+      // Make a copy of `_pendingEvents` to iterate through.
+      final List<TimelineEvent> events =
+          List<TimelineEvent>.from(_pendingEvents);
+
+      for (TimelineEvent event in events) {
+        final bool eventAdded = _maybeAddEventToFrame(event, frame);
+        if (eventAdded) _pendingEvents.remove(event);
+      }
     }
   }
 
-  void _handleAsyncInstantEvent(TraceEvent event) {
-    final String asyncUID = event.asyncUID;
+  /// Add event to an available frame in [_pendingFrames] if we can, or otherwise add it
+  /// to [_pendingEvents].
+  void _maybeAddEvent(TimelineEvent event) {
+    if (!event.isPipelineProduceFlow && !event.isPipelineConsumeFlow) {
+      // We do not care about other events.
+      return;
+    }
 
-    final TimelineEvent e = TimelineEvent(
-      event.name,
-      event.timestampMicros,
-      _getTimelineEventType(event),
-    );
-    e.endTime = event.timestampMicros + event.duration;
+    bool eventAdded = false;
 
-    final TimelineEvent parent = _asyncEvents[asyncUID];
-    if (parent != null) {
-      e.parent = parent;
+    final List<TimelineFrame> frames = _getAndSortWellFormedFrames();
+    for (TimelineFrame frame in frames) {
+      eventAdded = _maybeAddEventToFrame(event, frame);
     }
+
+    if (!eventAdded) _pendingEvents.add(event);
   }
 
-  void _handleAsyncEndEvent(TraceEvent event) {
-    final String asyncUID = event.asyncUID;
+  /// Add event `e` to frame `f` if it meets the necessary criteria.
```

Comment:
replace `e` with [e] and `f` with [f] so dartdoc does the right thing here and elsewhere.

Reply:
It seems dart doc doesn't actually do anything unless the vars are class members, but I will change to [ ] anyway for consistency.